### ISSUE TITLE
[AD] SCORPIO-Interface: Automatic registration of dimensions as variables

### ISSUE THREAD

**Issue by AaronDonahue** [E3SM-Project/scream#786]
netCDF makes a distinction between dimensions and variables.  All variables must be defined as having 1 or more dimensions, as they are registered with a particular netCDF file.  In other words, dimensions are registered first, and a variable cannot be registered if it has a dimension unknown to the netCDF file.

In the interface, this is handled via register_dimension and register_variable.

Note, that dimensions don't necessarily need to be written as variables too, however, it makes sense that they would be since that would allow users to understand exactly where/when data exists.

This task is to improve the current infrastructure to automatically add any new dimension as also a variable to be written.  Currently, the dimensions themselves need to be included as output fields, which is a bit cumbersome, and also opens the door to a user accidentally not including, say lat/lon, as output variables and getting frustrated when they are not present in the netCDF output file.

Additionally, dimensions should not be dependent on time (except for of course the time dimension).  So they should only be written once, right at initialization.

**Comment by PeterCaldwell** [comment]
I agree that this is important. When I've faced a netcdf file in the past that doesn't contain the variable associated with its dimensions, it has been really confusing and sometimes even caused errors in the tools I was using to explore the netcdf file. I generally consider netcdf files without their dimension information to be broken/wrong/bad.

**Comment by bartgol** [comment]
So, to be clear, you suggest that if a variable has a dimension 'x', we should have a variable called 'x' in the file storing all the possible values for that dimension?

**Comment by PeterCaldwell** [comment]
@bartgol - yes to the first part, no to the last. Dimensions are things like "level" or "latitude" which are ordered lists of values rather than "all possible values" which sounds like a mathematical set without ordering. I envision this issue to result in code that just follows up any add_dimension(x) call with an add_variable(x) call.

**Comment by bartgol** [comment]
Yes, sorry, poor word choice. But bottom line is that yo want the variable 'z' to store the actual z-coordinate, right?

**Comment by AaronDonahue** [comment]
Yes.  And we want it to be that when say "latitude" is added as a dimension it is also added as a variable.

For clarification, in netCDF when you declare a dimension all you do is give it a name and a length.  But it doesn't automatically store all the possible values for that dimension.  That has to be done by also adding it as a variable.  I would like to have that happen automatically.

**Comment by PeterCaldwell** [comment]
Yes, I agree with Luca's "bottom line".

**Comment by AaronDonahue** [comment]
@bartgol , have we already satisfied this issue?

**Comment by bartgol** [comment]
We did this for the lev/ilev dimension (I think you did it). We do not have it for any other dimension. I don't think we need it for other dimensions though. Rad should actually be doing it for the bands; I mean, I added the possibility for rad to do it, I don't recall right now if it does. But either way, bottom line is that the IO layer should not be in charge to do this. It can add all the grid "geo data" as outputs (that's how rrtmgp can cause its sw/lw bands to be outputed), but that's about it.

**Comment by bartgol** [comment]
And what would you like the var dim8 to contain? The numbers 0-7? What's the name of that? And what would the var "ncol" contain? 

IMHO, dimensions are just that. You can have a variable that runs along that dimension which specifies some quantity that's strictly related to the dimension. Eg we do that for time (which I don't like, since the dimension should just be "time_index")

But I don't see it as a general thing.

**Comment by bartgol** [comment]
We can definitely allow users to register a variable that describe the "coordinate" of each dim entry. But I would not automatically register a variable for each dimension. Some dimensions are just "counters", and the only meaningful variable one could associate with them would have values 0,1,2,...,N, which is not useful at all, imho.

We can make the interface more "coordinate" oriented. So in the grid class, rather than making the user call geo data, we would expose a method like `set_dim_coordinates(const string& dim_name, const Field& coords)` or something like that, that hopeully conveys more meaning.
